reporter/reporter.go

```diff
@@ -102,40 +102,47 @@ func Monitor(ctx context.Context) {
 	}
 }
 
-// A line from the backtrace.
-type BacktraceLine struct {
-	PC   uintptr
-	File string
-	Line int
-}
-
-// Error wraps an error with additional information, like a backtrace,
+// Error wraps an error with additional information, like a stack trace,
 // contextual information, and an http request if provided.
 type Error struct {
 	// The error that was generated.
 	Err error
 
-	// The backtrace.
-	Backtrace []*BacktraceLine
-
 	// Any freeform contextual information about that error.
 	Context map[string]interface{}
 
 	// If provided, an http request that generated the error.
 	Request *http.Request
+
+	// This is private so that it can be exposed via StackTrace(),
+	// which implements the stackTracker interface.
+	stackTrace *errors.StackTrace
 }
 
-// Make error implement the error interface.
+// Make Error implement the error interface.
 func (e *Error) Error() string {
 	return e.Err.Error()
 }
 
-// NewError wraps err as an Error and generates a backtrace pointing at the
+// Make Error implement the causer interface.
+func (e *Error) Cause() error {
+	return errors.Cause(e.Err)
+}
+
+// Make Error implement the stackTracer interface.
+func (e *Error) StackTrace() errors.StackTrace {
+	if e.stackTrace != nil {
+		return *e.stackTrace
+	}
+	return nil
+}
+
+// NewError wraps err as an Error and generates a stack trace pointing at the
 // caller of this function.
 func NewError(err error, skip int) *Error {
```

Comment:
So basically, if you wrap your errors correctly in your code, and reporter catches it at the top, it'll wrap it with `NewError`, which finds the best stacktrace. Then hb and nr reporters can just rely on the stacktracker and causer interfaces.

I think this is pretty good. I wonder if we may want the logic used by `reporter.NewError` exposed from our `errors` package. Can't come up with an example, but it feels like it belongs in the errors package to me.